karoo_gp/base_class.py

```diff
@@ -14,182 +14,165 @@
+        self.datetime = datetime.now().strftime('%Y-%m-%d_%H-%M-%S-%f')
+        self.filename = filename             # prefix for output files
+        self.output_dir = output_dir         # path to output directory
+        self.loader = DataLoader(self)       # set path, initialize logs
+        self.mode = mode                     # determines if pauses after fit
+        self.display = display               # determines when pause/log called
+        self.pause_callback = pause_callback # called throughout based on disp
+        self.history = {}                    # best score for each field/gen
+
+        # Initialize Population
+        self.functions = Functions(functions)
+        self.terminals = Terminals(terminals)
+        self.tree_type = tree_type           # (f)ull, (g)row or (r)amped 50/50
+        self.tree_depth_base = tree_depth_base # depth of initial population
+        self.tree_pop_max = tree_pop_max     # number of trees per generation
         self.population = Population.generate(
-            log=self.log, pause=self.pause, error=self.error,
-            tree_type=self.tree_type, tree_depth_base=self.tree_depth_base,
-            tree_depth_max=tree_depth_max, tree_pop_max=tree_pop_max,
-            functions=self.functions, terminals=self.terminals, rng=self.rng,
-            fitness_type=self.fitness_type
+            model=self,
+            functions=self.functions,
+            terminals=self.terminals,
+            tree_type=self.tree_type,
+            tree_depth_base=self.tree_depth_base,
+            tree_pop_max=tree_pop_max,
         )
-
         self.log(f'\n We have constructed the first, stochastic population of'
-                 f'{self.tree_pop_max} Trees'
-                 f'\n Evaluate the first generation of Trees ...')
-
-        if self.kernel == 'p':
-            self.fx_data_tree_write(self.population.trees, 'a')
-            sys.exit()
-
-        self.population.evaluate(
-            log=self.log, pause=self.pause, error=self.error,
-            data_train=self.data_train, kernel=self.kernel,
-            data_train_rows=self.data_train_rows,
-            tf_device_log=self.tf_device_log, class_labels=self.class_labels,
-            tf_device=self.tf_device, terminals=self.terminals,
-            precision=self.precision, savefile=self.savefile,
-            fx_data_tree_write=self.fx_data_tree_write
-        )
-
+                 f'{self.tree_pop_max} Trees. \n Evaluate the first generation'
+                 f' of Trees ...')
+        # if self.kernel == 'p':
+        #     self.fx_data_tree_write(self.population.trees, 'a')
+        #     sys.exit()
+
+        # Engine
+        if engine_type in ['numpy', 'np']:
+            self.engine = NumpyEngine(self)
+        elif engine_type in ['tensorflow', 'tf']:
+            self.engine = TensorflowEngine(self, tf_device, tf_device_log)
+        else:
+            raise ValueError(f'Unrecognized engine_type: {engine_type}')
```

Comment:
Actually, this might be a better way to do what `LabelEncoder` is doing in `MultiClassifierGP`. 

Right now `BaseGP` has a dummy variable `.encoder` that's called if set externally. 

Instead, the `engine_type` kwarg could be changed to `engine` and accept either a string ('numpy', 'tensorflow') or an instance of Engine. The `MultiCl...` can pass a custom engine like the one above that transforms the predictions directly, instead of using the encoder.

Reply:
More on this: The Keras [model.compile()](https://keras.io/api/models/model_training_apis/) API has a handful of fields that accept some combination of strings, classes, lists of either/both, or dicts of `{label: class}` pairs. This is what I aimed to emulate with `scoring`, `engine`, `terminals` and `functions`.

Reply:
Being able to pass strings is convenient, but limited to the strings that the function knows.  Passing `Engine` (sub)classes gives you more flexibility, but is a bit more involved since you have to first import the class.  Accepting both is generally something that should be avoided, but if this is common in this context, then it's probably ok -- after all "practicality beats purity".